### PR TITLE
Solution space handler

### PR DESCRIPTION
**Pull request by nbirillo**

### CODE REVIEW

File: src/main/solution_space/data_classes.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+from datetime import datetime
+
+from typing import List, Union
+from src.main.util import consts
+from src.main.canonicalization.canonicalization import get_code_from_tree
+from src.main.util.consts import EXPERIENCE, DEFAULT_VALUES, ACTIVITY_TRACKER_EVENTS
+
+
+class AtiItem:
+    # Todo: Add event_type to EVENT_EVENTS??
```
Comment (elena-lyulina) [inline src/main/solution_space/data_classes.py]: I guess, this todo is redundant now, we can remove it then

File: src/main/solution_space/solution_space_handler.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    return Code(ast=tree, rate=rate)
+
+
+def __convert_to_datetime(df: pd.DataFrame) -> None:
+    for column in [consts.CODE_TRACKER_COLUMN.DATE, consts.ACTIVITY_TRACKER_COLUMN.TIMESTAMP_ATI]:
+        df[column.value] = pd.to_datetime(df[column.value], errors='ignore')
+
+
+def __add_user_solutions(file: str, task: str) -> List[Tuple[Code, CodeInfo]]:
+    log.info(f'Start solution space creating file {file} for task {task}')
```
Comment (elena-lyulina) [inline src/main/solution_space/solution_space_handler.py]: creating for file :)

File: src/main/solution_space/solution_space_handler.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    task_index = __get_task_index(task)
+    i, code_info_chain = 0, []
+    user = __get_user(solutions)
+    while i < solutions.shape[0]:
+        old_index = i
+        i, ati_actions, tree = __find_same_fragments(solutions, i)
+        code = __get_code(solutions, old_index, task_index, tree)
+        code_info = __get_code_info(solutions, user, old_index, ati_actions)
+        code_info_chain.append((code, code_info))
+    log.info(f'Finish solution space creating file {file} for task {task}')
```
Comment (elena-lyulina) [inline src/main/solution_space/solution_space_handler.py]: creating for file :)

File: src/test/solution_space/solution_space_handler/filter_incorrect_fragments_tests.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    'Data': [1, 2, 3],
+    CODE_TRACKER_COLUMN.TESTS_RESULTS.value: [LESS_TASKS, LESS_TASKS, LESS_TASKS]
+}
+
+
+def run_test(before_filtering: pd.DataFrame, expected_after_filtering: pd.DataFrame) -> bool:
+    try:
+        actual_after_filtering = __filter_incorrect_fragments(before_filtering)
+        actual_after_filtering.index = [*range(actual_after_filtering.shape[0])]
+    except ValueError:
```
[the remark this answers is not in the document]
Comment (nbirillo) [reply src/test/solution_space/solution_space_handler/filter_incorrect_fragments_tests.py]: Thanks! I have never used it. It looks well

**Pull request merged**